components/widgets/inputs/users_emails_input.jsx

```diff
@@ -0,0 +1,212 @@
+        validAddressMessageDefault: 'Add **{email}**',
+        loadingMessageId: t('widgets.users_emails_input.loading'),
+        loadingMessageDefault: 'Loading',
+    };
+
+    loadingMessage = () => {
+        if (!this.context.intl) {
+            return 'Loading';
+        }
+
+        return this.context.intl.formatMessage({
+            id: this.props.loadingMessageId,
+            defaultMessage: this.props.loadingMessageDefault,
+        });
+    }
+
+    getOptionValue = (user) => {
+        return user.id || user.value;
+    }
+
+    formatUserName = (user) => {
+        let displayName = '@' + user.username + ' - ' + user.first_name + ' ' + user.last_name;
+        if (user.nickname) {
+            displayName = displayName + ' (' + user.nickname + ')';
+        }
+        return displayName;
+    }
+
+    formatShortUserName = (user) => {
+        if (user.first_name || user.last_name) {
+            return user.first_name + ' ' + user.last_name;
+        }
+        return user.username;
+    }
+
+    formatOptionLabel = (user, options) => {
+        const profileImg = imageURLForUser(user);
+        const avatar = (
+            <img
+                className='avatar'
+                alt={`${user.username || 'user'} profile image`}
+                src={profileImg}
+            />
+        );
+
+        if (options.context === 'menu') {
+            if (user.value && isEmail(user.value)) {
+                return this.getCreateLabel(user.value);
+            }
+            return (
+                <React.Fragment>
+                    {avatar}
+                    {this.formatUserName(user)}
+                </React.Fragment>
+            );
+        }
+
+        if (user.value && isEmail(user.value)) {
+            return (
+                <React.Fragment>
+                    <MailIcon className='mail-icon'/>
+                    <span>{user.value}</span>
+                </React.Fragment>
+            );
+        }
+
+        return (
+            <React.Fragment>
+                {avatar}
+                {this.formatShortUserName(user)}
+            </React.Fragment>
+        );
+    }
+
+    onChange = (value) => {
+        if (this.props.onChange) {
+            this.props.onChange(value.map((v) => {
+                if (v.id) {
+                    return v;
+                }
+                return v.value;
+            }));
+        }
+    }
+
+    getCreateLabel = (value) => (
+        <React.Fragment>
+            <MailPlusIcon className='mail-plus-icon'/>
+            <FormattedMarkdownMessage
+                key='widgets.users_emails_input.valid_email'
+                id={this.props.validAddressMessageId}
+                defaultMessage={this.props.validAddressMessageDefault}
+                values={{email: value}}
+            />
+        </React.Fragment>
+    );
+
+    NoOptionsMessage = (props) => {
+        const inputValue = props.selectProps.inputValue;
+        if (inputValue) {
```

Comment:
That's not exactly what I was going for, I was thinking on the fact that we have to almost-identical component returns, so we could put those messages in a variable and have one return with one component:

```js
let messageId = this.props.noOptionsMessageId || t('widgets.users_emails_input.empty');
let defaultMessage = this.props.noOptionsMessageDefault || 'Add users or email addresses';

if (inputValue) {
  let messageId = this.props.noMatchMessageId || t('widgets.users_emails_input.no_user_found_matching');
  let defaultMessage = this.props.noMatchMessageDefault || 'No one found matching **{text}**, type email address';
}

return (
  <div className='users-emails-input__option'>
      <FormattedMarkdownMessage
          id={messageId}
          defaultMessage={defaultMessage}
          values={{text: inputValue}}
      >
          {(message) => (
              <components.NoOptionsMessage {...props}>
                  {message}
              </components.NoOptionsMessage>
          )}
      </FormattedMarkdownMessage>
  </div>
);
```

Reply:
Good point, I'll do that too

Reply:
Done! :)

Reply:
:D